### DISCUSSION TITLE
What's the future of AWS Copilot?

### DISCUSSION THREAD

**Discussion by afgallo** [aws/copilot-cli#5925] [General]
Hi All,

I am not sure if it is just me but AWS Copilot doesn't seem to have a lot of traction these days.

This tool used to have a pretty good monthly release cycle and now it's been months since a release that was packed with new features such as https://github.com/aws/copilot-cli/releases/tag/v1.33.0.

Is this a tool that AWS is still investing resources at? Or is this simply a product that is considered "done"?

What can we expect from future improvements given the Roadmaps don't seem to be updated very often?

I run a few platforms in Production where we use Copilot and I was just wondering if I should be concerned at all.

Cheers,
Andre

**Comment by iamhopaul123** [comment]
Hello @afgallo. We are not actively developing any new features for Copilot but we still do maintenance work. Copilot is a client side tooling so ideally if you use any available version it should be stable enough for your prod application. There could be scenarios that a new dependency vulnerabilities has been uncovered for existing version and in that case we'll still fix any high severity security vulnerabilities and release ad hoc. Does that address your concern?

**Reply by afgallo** [reply]
Thanks @iamhopaul123. It's sad to see that there will be no further new development done in Copilot. I feel like there is a bunch of new features that could be added to make things even easier.

And yes, it addresses my question so my take away is that the tool is in maintenance only.

Cheers

**Reply by rdig-jhayes** [reply]
Hi @iamhopaul123 can you clarify what it means that your team is still doing maintenance work? Are you maintaining the existing Copilot functionality (keeping up with changes to ECS and other services it uses)?

Put another way: if a change to another AWS service impacts Copilot, will you be fixing that, or will Copilot's functionality degrade from now on? Thank you.

**Comment by PedroAlvarado** [comment]
@iamhopaul123 It's important to make it clear that AWS Copilot is in maintenance mode, as this isn't obvious to newcomers. I suggest highlighting this prominently on the GitHub repo and website for the benefit of current and future users.

**Comment by hiroki-gota** [comment]
Agreed with others. We have 2 production workloads using CoPilot. We're also intending to use it for the next project and we're about to create a new instance. What is the official recommendation @iamhopaul123 for a new project? Is there an alternative recommended tool? including CDK?

**Reply by xinye2005** [reply]
Agreed. If AWS really decides to stop this project, they should offer alternatives. But in fact, AWS copilot helped ECS generate a lot of revenue by implementing the reference architecture. I really hope they can continue maintaining this.

**Comment by dfuentes77** [comment]
We should also point out the obvious to many of us, that there is a huge amount of documentation inside and outside of AWS docs, articles, blogs, instructions, that's still in place.  The external stuff you can't change much but so much internal documentation (even in the AWS console) still points to AWS copilot.  And if this project is officially in maintenance mode, all that documentation should have already changed, since that's leading new and current AWS users to an unexpected dead-end.

**Comment by gj02ib65** [comment]
I am supporting a customer that has used Copilot to deploy ECS solutions in 11 AWS accounts.  We deployed this solution with CodeCommit and pipelines and it was great, but CodeCommit has been abandoned by AWS and now AWS Copilot.  

I appreciate that you can say that Copilot it is functionally complete.  I am amazed with the ability and adaptability to different deployment strategies, but I would like to see more commitment in the maintenance mode to maintain the functionality of Copilot as the underling services of AWS changes.  

As AWS implements changes to services, I am starting to see cracks emerge in Copilot deployments.  Yes, you can find ways to work around those, but I would like to see AWS continue to fund resources to fix those items one time in Copilot, rather than each customer needing to troubleshoot and patch around those issues. 

@iamhopaul123 I would like to see an expanded definition of maintenance to include maintaining functionality of Copilot.  I understand you may need to deprecate the ability of using CodeCommit for instance and that no new features will be added.  But I would like to see continued support and maintenance of features of AWS Copilot by adapting to the changing AWS infrastructure.

**Reply by rdig-jhayes** [reply]
I also have Copilot-based dev processes and I think @gj02ib65 's point is worth repeating:

Maintenance mode could mean (a) continued commits to support changes made to the AWS services which Copilot interacts with, or (b) it could mean security patches only.

The former option would allow Copilot to remain a useful tool for those who have adopted it. The latter would render Copilot increasingly broken for users as certain AWS services evolve. If this latter option is the path forward, @dfuentes77 is correct, the documentation around Copilot should change ASAP to encourage people simply not to use it, since any change to the interface for an AWS service with which Copilot interacts could make it immediately unusable.

**Reply by afgallo** [reply]
Had to dig this out of commits history, this is officially unmaintained ‼
https://github.com/aws/copilot-cli/pull/5975/files

**Reply by rdig-jhayes** [reply]
The fact that [the blog post](https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-copilot-cli-effective-february-3-2025/) referenced in [that commit & PR](https://github.com/aws/copilot-cli/pull/5974) was deleted and the PR's changes [reverted 2 hours after merging](https://github.com/aws/copilot-cli/pull/5975) makes that announcement **unofficial**. It's a very confusing situation since November 4.

**Reply by afgallo** [reply]
True @rdig-jhayes. AWS has failed to communicate this clearly. However, I do think there is enough evidence that this will no longer be a viable solution. For those interested in sharing their experiences and what they want to do in the future, there are some discussions happening here: https://github.com/aws/copilot-cli/pull/5975

**Comment by jasonforte** [comment]
It seems we have an official answer from AWS, EOL is 12 June 2026: https://aws.amazon.com/blogs/containers/announcing-the-end-of-support-for-the-aws-copilot-cli/

How's everyone finding the alternatives mentioned? Has anyone tried ECS Express Mode for anything beyond a basic setup?

**Reply by duplxey** [reply]
I'm currently researching the mentioned alternatives. It seems like they're all low-level and only partially solve what Copilot CLI managed. 

- ECS Express Mode seems great for deploying ECS containers, but that's where it ends. There's no environment management, image management (building & pushing), nice environmental variable/secrets injections, databases, addons, etc.
- AWS CloudFormation and AWS CDK on the other hand, aren't really alternatives. They're just IaC that you can use to replicate the Copilot CLI behaviour. I thought the whole point of Copilot was to not have to deal with CFN templates/CDK.

I guess the way to go is to implement everything by yourself using CFN/CDK and maybe utilize ECS Express Mode to simplify the ECS deployment. If anyone has a better idea, please let me know.

**Reply by CharlieDigital** [reply]
Very sad to hear that they are sunsetting this because this is genuinely the best way to work on the AWS stack.

The tooling took a lot of bespoke code and complexity out of the process and made it really easy to fully leverage the platform capabilities.  Very surprised that they are killing this because my sense is that the small investment in this CLI tooling makes AWS infrastructure much more accessible and easier to use.
